<p style="text-align: center; font-size: 24px; font-weight: bold;">
PROJECT: Optimizing Oil Well Placement: A Predictive Analysis of Profits and Risks for OilyGiant Mining Company
</p>



#### Introduction.  
In this project, we will work with data from the OilyGiant mining company to identify the best location for new oil wells. The goal is to predict the volume of reserves and evaluate potential profit margins for three selected regions. To achieve this, we will analyze key oil well parameters, including oil quality and volume of reserves, for each region. First, we will preprocess and clean the available data to ensure its readiness for analysis. We will then build a Linear Regression model to predict the volume of reserves for new wells. After identifying the wells with the highest predicted reserves, we will calculate the total profit for each region based on these predictions. To assess the risks, we will apply the Bootstrapping technique to estimate the potential profits and risks of losses for each region. Our final task is to recommend the region with the highest expected profit and the lowest risk, ensuring that the risk of losses does not exceed 2.5%. The findings will provide insights into the best locations for the development of new oil wells. 

#### Importing functions

In [4]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
import numpy as np


### Data preparation

#### Data preprocessing

In [7]:
# Here we download three datasets
data_0 = pd.read_csv('geo_data_0.csv')
data_1 = pd.read_csv('geo_data_1.csv')
data_2 = pd.read_csv('geo_data_2.csv')


In [8]:
# Data preprocessing
print(data_0.info())
print(data_0.head())
print(data_0.isna().sum())
print(data_0.duplicated().sum())
print(data_0['id'].nunique())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
0
99990


In [9]:
# Data preprocessing
print(data_1.info())
print(data_1.head())
print(data_1.isna().sum())
data_1.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64


0

In [10]:
# Data preprocessing
print(data_2.info())
print(data_2.head())
print(data_2.isna().sum())
data_2.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64


0

All three datasets have no missing and duplicated values. 

#### Data cleaning

In [13]:
#All three datasets have ['id'] feature (cathegorical datatype). The values of ['id'] cannot determine 
#the quality of oil and product volume. Therefore we remove this feature from all three datasets. 

data_0 = data_0.drop(['id'], axis=1)
data_1 = data_1.drop(['id'], axis=1)
data_2 = data_2.drop(['id'], axis=1)

print(data_0.head(5))
print(data_1.head(5))
print(data_2.head(5))


         f0        f1        f2     product
0  0.705745 -0.497823  1.221170  105.280062
1  1.334711 -0.340164  4.365080   73.037750
2  1.022732  0.151990  1.419926   85.265647
3 -0.032172  0.139033  2.978566  168.620776
4  1.988431  0.155413  4.751769  154.036647
          f0         f1        f2     product
0 -15.001348  -8.276000 -0.005876    3.179103
1  14.272088  -3.475083  0.999183   26.953261
2   6.263187  -5.948386  5.001160  134.766305
3 -13.081196 -11.506057  4.999415  137.945408
4  12.702195  -8.147433  5.004363  134.766305
         f0        f1        f2     product
0 -1.146987  0.963328 -0.828965   27.758673
1  0.262778  0.269839 -2.530187   56.069697
2  0.194587  0.289035 -5.586433   62.871910
3  2.236060 -0.553760  0.930038  114.572842
4 -0.515993  1.716266  5.899011  149.600746


### LinearRegression for model training 

In [15]:
# data_0

features_0 = data_0.drop(['product'], axis=1) 
target_0 = data_0['product']

features_train_0, features_valid_0, target_train_0, target_valid_0 = train_test_split(features_0, target_0, test_size=25000, random_state=12345)

model = LinearRegression()
model.fit(features_train_0, target_train_0)
predict_0 = model.predict(features_valid_0)

predict_0_avg = predict_0.mean()
print(f'Avrage volume of predicted reserves = {predict_0_avg:.4f}')


# below, we calculate RMSE, R2 and MAE metrics:
mse_0 = mean_squared_error(target_valid_0, predict_0)
R2_0 = model.score(features_valid_0, target_valid_0)

print (f'RMSE_0 = {mse_0**0.5:.4f}')
print(f'R2_0 metric value: {R2_0:.4f}')
print(f'MAE_0= {mean_absolute_error(target_valid_0, predict_0):.4f}')
                       

Avrage volume of predicted reserves = 92.5926
RMSE_0 = 37.5794
R2_0 metric value: 0.2799
MAE_0= 30.9196


In [16]:
# data_1
features_1 = data_1.drop(['product'], axis=1) 
target_1 = data_1['product']

features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split(features_1, target_1, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(features_train_1, target_train_1)
predict_1 = model.predict(features_valid_1)

predict_1_avg = predict_1.mean()
print(f'Average volume of predicted reserves = {predict_1_avg:.4f}')

# below, we calculate RMSE, R2 and MAE metrics:
mse_1 = mean_squared_error(target_valid_1, predict_1)
R2_1 = model.score(features_valid_1, target_valid_1)

print (f'RMSE_1 = {mse_1**0.5:.4f}')
print(f'R2_1 metric value: {R2_1:.4f}')
print(f'MAE_1= {mean_absolute_error(target_valid_1, predict_1):.4f}')


Average volume of predicted reserves = 68.7285
RMSE_1 = 0.8931
R2_1 metric value: 0.9996
MAE_1= 0.7188


In [17]:
# data_2

features_2 = data_2.drop(['product'], axis=1) 
target_2 = data_2['product']

features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split(features_2, target_2, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(features_train_2, target_train_2)
predict_2 = model.predict(features_valid_2)

predict_2_avg = predict_2.mean()
print(f'Avrage volume of predicted reserves = {predict_2_avg:.4f}')

# below, we calculate RMSE, R2 and MAE metrics:
mse_2 = mean_squared_error(target_valid_2, predict_2)
R2_2 = model.score(features_valid_2, target_valid_2)

print (f'RMSE_2 = {mse_2**0.5:.4f}')
print(f'R2_2 metric value: {R2_2:.4f}')
print(f'MAE_2= {mean_absolute_error(target_valid_2, predict_2):.4f}')


Avrage volume of predicted reserves = 94.9650
RMSE_2 = 40.0297
R2_2 metric value: 0.2052
MAE_2= 32.7927


##### LinearRegression model predictions give the following average volume of predicted reserves: 

1) data_0 (region_1) -  92.5926; 
2) data_1 (region_2)  - 68.7285  
3) data_2 (region_3) - 94.9650





### Profit and risk of losses calculations

In [20]:
# given data for for profit calculation

budget_for_200_wells = 100000000
number_of_wells = 200
cost_per_well = budget_for_200_wells/number_of_wells
revenue_per_unit = 4500

def profit_calculation(predictions, targets, cost_per_well, revenue_per_unit, count):
    predict_sorted = predictions.sort_values(ascending=False)
    selected_targets = targets[predict_sorted.index][:count]
    
    # Calculate total revenue and profit
    total_revenue = selected_targets.sum() * revenue_per_unit
    total_cost = cost_per_well * count
    profit = total_revenue - total_cost
    
    return profit


# Here, we calculate the volume of reserves sufficient for developing a new well without losses. 

required_budget_per_well = budget_for_200_wells/number_of_wells

required_volume = required_budget_per_well/revenue_per_unit
                              
print(f'Required_budget_per_well = {required_budget_per_well:.2f}')
print(f'Required_volume = {required_volume:.2f}')       #unit is thousand barrels


Required_budget_per_well = 500000.00
Required_volume = 111.11


##### The calculated volume of reserves sufficient for developing a new well without losses is 111.11 thousand barrels. This value is larger than the above shown predected average values of the regions (1) region_1 -  92.5926 thousand barrels, region_2  - 68.7285 thousand barrels, region_3 - 94.9650 thousand barrels. Therefore, below we will evaluate top 200 wells for each region.

In [22]:
# functions for calculating the profit metrics


def profit_metrics(predictions, targets, cost_per_well, revenue_per_unit, count, num_samples=1000):
    
    state = np.random.RandomState(12345)
    
    profit_subsample = []
    
    for _ in range(num_samples):
        target_subsample = targets.sample(n=500, replace=True, random_state=state)
        predict_subsample = predictions[target_subsample.index]
        profit_subsample.append(profit_calculation(predict_subsample, target_subsample, cost_per_well, revenue_per_unit, count))
    
    profit_subsample = pd.Series(profit_subsample)

    lower = profit_subsample.quantile(0.025)
    upper = profit_subsample.quantile(0.975)
    profit_mean = profit_subsample.mean()
    risk_of_loss = (profit_subsample < 0).mean() * 100
    
    return profit_mean, lower, upper, risk_of_loss


In [23]:
# data_0: functions for calculating the profit metrics

#creating df_0 by combining the target_valid_0 and predict_0

predictions_0 = pd.Series(predict_0, index = target_valid_0.index)
df_0 = pd.DataFrame({'predictions_0':predictions_0, 'target_valid_0':target_valid_0})

predictions_0 = df_0['predictions_0']
target_valid_0 = df_0['target_valid_0']


profit_mean, lower, upper, risk_of_loss = profit_metrics(predictions_0, target_valid_0, cost_per_well, revenue_per_unit, 200)

print(f'Average profit_0 = {profit_mean:.2f}')
print(f'95% confidence interval_0: lower = {lower:.2f},  upper = {upper:.2f}')
print(f'Risk of loss_0 = {risk_of_loss:.2f}%')


Average profit_0 = 4259385.27
95% confidence interval_0: lower = -1020900.95,  upper = 9479763.53
Risk of loss_0 = 6.00%


In [24]:
# data_1: functions for calculating the profit metrics

#creating df_1 by combining the target_valid_1 and predict_1

predictions_1 = pd.Series(predict_1, index = target_valid_1.index)
df_1 = pd.DataFrame({'predictions_1':predictions_1, 'target_valid_1':target_valid_1})

predictions_1 = df_1['predictions_1']
target_valid_1 = df_1['target_valid_1']


profit_mean, lower, upper, risk_of_loss = profit_metrics(predictions_1, target_valid_1, cost_per_well, revenue_per_unit, 200)

print(f'Average profit_1 = {profit_mean:.2f}')
print(f'95% confidence interval_1: lower = {lower:.2f},  upper = {upper:.2f}')
print(f'Risk of loss_1 = {risk_of_loss:.2f}%')


Average profit_1 = 5152227.73
95% confidence interval_1: lower = 688732.25,  upper = 9315475.91
Risk of loss_1 = 1.00%


In [25]:
# data_2: functions for calculating the profit metrics

#creating df_2 by combining the target_valid_2 and predict_2

predictions_2 = pd.Series(predict_2, index = target_valid_2.index)
df_2 = pd.DataFrame({'predictions_2':predictions_2, 'target_valid_2':target_valid_2})

predictions_2 = df_2['predictions_2']
target_valid_2 = df_2['target_valid_2']


profit_mean, lower, upper, risk_of_loss = profit_metrics(predictions_2, target_valid_2, cost_per_well, revenue_per_unit, 200)

print(f'Average profit_2 = {profit_mean:.2f}')
print(f'95% confidence interval_2: lower = {lower:.2f},  upper = {upper:.2f}')
print(f'Risk of loss_2 = {risk_of_loss:.2f}%')


Average profit_2 = 4350083.63
95% confidence interval_2: lower = -1288805.47,  upper = 9697069.54
Risk of loss_2 = 6.40%


#### Profit and risk of losses calculations show the following results: 

data_0:  
Average profit_0 = 4259385.27     
95% confidence interval: lower_0 = -1020900.95,  upper_0 = 9479763.53
Risk of loss_0 = 6.00%

data_1: 
Average profit_1 = 5152227.73    
95% confidence interval: lower_1 = 688732.25, upper_1 = 9315475.91
Risk of loss_1 = 1.00%

data_2: 
Average profit_2 = 4350083.63    
95% confidence interval: lower_2 = -1288805.47,   upper_2 = 9697069.54
Risk of loss_2 = 6.40%


### Conclusions. 
In this project, we analyzed data from the OilyGiant mining company to determine the best location for developing new oil wells. Using data on oil well parameters—specifically oil quality and volume of reserves—we built a Linear Regression model to predict the reserves in potential new wells. After selecting the wells with the highest predicted reserves, we estimated the total profits for each region. We then applied the Bootstrapping technique to calculate both the average profits and the risks of losses for each region. Our analysis revealed the following results:

Region 1: Average profit = 4,259,385.27, Risk of losses = 6%
Region 2: Average profit = 5,152,227.73, Risk of losses = 1%
Region 3: Average profit = 4,259,385.27, Risk of losses = 6%
Based on these findings, we recommend Region 2 as the optimal location for new oil wells, as it meets the project requirement of having a risk of losses below 2.5%. Region 2 offers the highest total profit and the lowest risk, making it the best choice for development.